Specify EmailSender interface

selo-ecube opened the pull request:
The interface includes a proposal for a default method for sending e-mails synchronously, e.g. in the tests.

selo-ecube committed:
Specify EmailSender interface

The interface includes a proposal for a default method for sending e-mails synchronously, e.g. in the tests.

lauraluiz commented:
@selo-ecube is it ready to review? If it is, don't forget to mention @lauraluiz and @schleichardt so we start reviewing it :)

selo-ecube committed:
Pass MessageAuthor to EmailSender.send instead of MimeMessage

Because the MimeMessage constructor requires a Session, clients of the EmailSender
cannot use it. Instead, the EmailSender creates a MimeMessage and passes it to
the MessageAuthor that is passed to EmailSender.send(...).

selo-ecube commented:
Ah, sorry, yes, when the changes have passed travis, they are ready for review, @lauraluiz and @schleichardt .
I noticed that MimeMessage has no no-arg constructor and adjusted the interface. The interface also mentions async instead of lazy sending now.

schleichardt commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -1,6 +1,82 @@
+     * The {@link CompletionStage} returned by this method may be combined with further completion stages. To directly
+     * send the email and get its message ID, obtain a completable future as in the following example.
+     * <pre>{@code
+     * String messageID = completionStage.toCompletableFuture().get();
```
I would prefer `join` here since it is not throwing checked exceptions.

schleichardt commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -1,6 +1,82 @@
+    default
+    @Nonnull
+    String sendAndWait(@Nonnull final MessageAuthor messageAuthor, final int timeout, @Nonnull final TimeUnit unit)
+            throws ExecutionException, TimeoutException, InterruptedException {
```
here I would also prefer to only receive runtime exceptions.
It would be okay for now to catch those exceptions and may wrap them or the causse of ExecutionException in a CompletionException.

schleichardt replied:
especially in functional programming with Java 8 there is no place for checked exceptions in the default lambdas.

lauraluiz replied:
I would remove this synchronous method. In all the other interfaces it's never provided either. I will find a way to maybe provide something so that people use the correct way to execute things synchronously, but not from the interface.

schleichardt commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSenderException.java:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @see Exception#getCause()
+ */
+public class EmailSenderException extends Exception {
```
I would really prefer a RuntimeException.

schleichardt commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -1,6 +1,82 @@
+     * <a href="http://javamail.java.net/nonav/docs/JavaMail-1.5.pdf">Java Mail Specification</a> on how to configure
+     * instances of {@link MimeMessage}. The following code shows a simple example.
+     * <pre>{@code
+     * CompletionStage<String> completionStage = emailSender.send(msg -> {
```
this is necessary because of the construction of the mail already requires the configuration?

schleichardt replied:
http://docs.oracle.com/javaee/6/api/javax/mail/internet/MimeMessage.html#constructor_summary

Clever solution!

schleichardt commented on email-common/src/test/java/EmailSenderTest.java:
```diff
@@ -0,0 +1,42 @@
+        final int timeout = 1000;
+        final TimeUnit timeUnit = TimeUnit.SECONDS;
+        final CompletableFuture completableFuture = mock(CompletableFuture.class);
+        when(completableFuture.get(timeout, timeUnit)).thenReturn(ID);
```
_all_ the mockito stuff could be avoided by using

```
       final EmailSender sender = new EmailSender() {
             @Nonnull
             @Override
             public CompletionStage<String> send(@Nonnull final MessageAuthor author) {
                 return CompletableFuture.completedFuture(ID);
             }
         };
```

lauraluiz replied:
You are being pessimist @schleichardt :D

```
final EmailSender sender = a -> CompletableFuture.completedFuture(ID);
```

It's all it takes. That's why I insisted on the `@FunctionalInterface`, to save a lot of code on tests.

schleichardt commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -1,6 +1,82 @@
+     * @see #sendAndWait(MessageAuthor, int, TimeUnit)
+     */
+    @Nonnull
+    CompletionStage<String> send(@Nonnull final MessageAuthor messageAuthor);
```
it also be directly

```
@Nonnull
CompletionStage<String> send(@Nonnull final Consumer<MimeMessage> messageAuthor);
```

Since MessageAuthor is a functional interface without default methods it is easier to understand the
send signature with inlining the type.

With reading MessageAuthor in the signature without knowing the type caused confusion since I want to send a mail and an author is just a part of it.

lauraluiz replied:
Yes I vote for this too. I was confused by the MessageAuthor class, which wouldn't have happened if I would have seen Consumer<MimeMessage> instead.

schleichardt commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -1,6 +1,82 @@
 
-    // TODO
+    /**
+     * Create a completion stage that asynchronously sends an e-mail using the configuration of this e-mail
```
You documentation skills are remarkable and an inspiration. Thanks!

schleichardt commented on email-smtp/src/main/java/io/commercetools/sunrise/email/smtp/SmtpAuthEmailSender.java:
```diff
@@ -1,8 +1,32 @@
+        return new SmtpAuthEmailSender(properties);
+    }
+
+    @Override
```
actually I would prefer that visibility specifiers and annotations are not mixed and the annotations comes first with every annotation in one line

schleichardt commented on email-smtp/src/main/java/io/commercetools/sunrise/email/smtp/SmtpAuthEmailSender.java:
```diff
@@ -1,8 +1,32 @@
+
+    public static
+    @Nonnull
+    SmtpAuthEmailSender of(@Nonnull final Properties properties) {
```
I would suggest to not use `of` methods in a class hierarchy like this, since it is not possible to be consistent with the subclasses. If `of(Properties)` is used here, it is not (nicely) possible for the Gmail subclass to declare an of method with the same signature.

schleichardt commented on build.sbt:
```diff
@@ -57,7 +67,8 @@ def configCommonTestSettings(scopes: String) = Seq(
     "org.assertj" % "assertj-core" % "3.4.1" % scopes,
-    "com.icegreen" % "greenmail" % "1.5.0" % scopes
+    "com.icegreen" % "greenmail" % "1.5.0" % scopes,
+    "org.mockito" % "mockito-core" % "1.10.19" % scopes
```
I actually dislike the usage of mockito and similar frameworks. If you don't use a mocking framework then you see faster API design problems and refactor until it is easy testable.

lauraluiz commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -1,6 +1,82 @@
+     * Create a completion stage that asynchronously sends an e-mail using the configuration of this e-mail
+     * service. Before this method returns, the message author argument of this method is passed an empty
+     * {@link MimeMessage} that the {@link EmailSender} created. The message author shall prepare the message
+     * so it can be send. Sending will happen asynchronously, though.
```
*so it can be sent.

selo-ecube committed:
Implement review comments for https://github.com/commercetools/commercetools-sunrise-java-email/pull/4

I also removed the timeout parameters from EmailSender.sendAndWait(...). Timeout handling
will happen uniformly in EmailSender.send(...) instead. SmtpAuthEmailSender will use the timeout
handling of the SMTP transport of the Java Mail API.

lauraluiz commented:
Reviewed 👍 I think @schleichardt finished too

selo-ecube committed:
Remove EmailSender.sendAndWait(...)

Because the method implementation is identical to sample code provided in the JavaDoc
of EmailSender.send(...).

selo-ecube committed:
Remove EmailSenderTest

Because it tested a default method that has been removed.

selo-ecube commented:
Thanks for the informative reviews, @lauraluiz and @schleichardt ! Your reviews are covered in the latest commit. Please review again!

schleichardt commented:
👍 😃

lauraluiz commented:
Green light here too! You can merge.
